src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+            Finalize();
+            return std::nullopt;
+        }
+    }
+    return std::nullopt;
+}
+
+bool HeadersSyncState::ValidateAndStoreHeadersCommitments(const std::vector<CBlockHeader>& headers)
+{
+    // The caller should not give us an empty set of headers.
+    Assume(headers.size() > 0);
+    if (headers.size() == 0) return true;
+
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return false;
+
+    if (m_last_header_received.IsNull()) {
+        m_last_header_received = m_chain_start->GetBlockHeader();
+        m_current_height = m_chain_start->nHeight;
+    }
+
+    if (headers[0].hashPrevBlock != m_last_header_received.GetHash()) {
+        // Doesn't connect to what we expect
+        Finalize();
+        return false;
+    }
+
+    // If it does connect, (minimally) validate and store a commitment to each one.
+    for (const auto& hdr : headers) {
+        if (!ValidateAndProcessSingleHeader(m_last_header_received, hdr, m_current_height+1)) {
+            Finalize();
+            return false;
+        }
+    }
+
+    if (m_current_chain_work >= m_minimum_required_work) {
+        m_blockhash_with_sufficient_work = headers.back().GetHash();
+        m_redownloaded_headers.clear();
+        m_redownload_buffer_last_height = m_chain_start->nHeight;
+        m_redownload_buffer_first_prev_hash = m_chain_start->GetBlockHash();
+        m_redownload_buffer_last_hash.SetNull();
+        m_download_state = State::REDOWNLOAD;
+    }
+    return true;
+}
+
+bool HeadersSyncState::ValidateAndProcessSingleHeader(const CBlockHeader& previous, const CBlockHeader& current, int64_t current_height)
+{
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return false;
+
+    // Verify that the difficulty isn't growing too fast; an adversary with
+    // limited hashing capability has a greater chance of producing a high
+    // work chain if they compress the work into as few blocks as possible,
+    // so don't let anyone give a chain that would violate the difficulty
+    // adjustment maximum.
+    if (!PermittedDifficultyTransition(m_consensus_params, current_height,
+                previous.nBits, current.nBits)) {
+        return false;
+    }
+
+    if (!CheckProofOfWork(current.GetHash(), current.nBits, m_consensus_params)) return false;
+
+    if ((current_height - m_chain_start->nHeight) % HEADER_COMMITMENT_FREQUENCY == 0) {
+        // Try to add a commitment.
+        m_header_commitments.push_back(m_hasher(current.GetHash()) & 1);
+        if (m_header_commitments.size() > m_max_commitments) {
+            // The peer's chain is too long; give up.
+            // TODO: disconnect this peer.
+            LogPrint(BCLog::NET, "headers chain is too long; giving up sync peer=%d\n", m_id);
+            return false;
+        }
+    }
+
+    m_current_chain_work += GetBlockProof(CBlockIndex(current));
```

Comment:
In commit "Utilize anti-DoS headers download strategy"

Just a thought for later, but it's rather ugly to have to construct a `CBlockIndex` object just to be able to call `GetBlockProof`. I think `GetBlockProof` should work (or have a variant that works) with a `CBlockHeader` too, or even just the nBits value.